Sync/Synchronizers/IndependentRecordSynchronizer.swift

```diff
@@ -120,4 +120,28 @@ extension TimestampedSingleCollectionSynchronizer {
                 return deferMaybe(timestamp)
             }
     }
+
+    func uploadRecordsSingleBatch<T>(_ records: [Record<T>], lastTimestamp: Timestamp, storageClient: Sync15CollectionClient<T>) -> Deferred<Maybe<(timestamp: Timestamp, succeeded: [GUID])>> {
+        if records.isEmpty {
+            log.debug("No modified records to upload.")
+            return deferMaybe((timestamp: lastTimestamp, succeeded: []))
+        }
+
+        var succeeded: [GUID] = []
+        func onCollectionUploaded(result: POSTResult, timestamp: Timestamp?) -> DeferredTimestamp {
+            let stats = SyncUploadStats(sent: result.success.count, sentFailed: result.failed.count)
+            self.statsSession.recordUpload(stats: stats)
+            succeeded = result.success
+            return deferMaybe(timestamp ?? lastTimestamp)
+        }
+
+        let batch = storageClient.newBatch(ifUnmodifiedSince: (lastTimestamp == 0) ? nil : lastTimestamp, onCollectionUploaded: onCollectionUploaded)
+        return batch.addRecords(records, singleBatch: true)
+            >>> batch.endSingleBatch
+            >>> {
+                let timestamp = batch.ifUnmodifiedSince ?? lastTimestamp
```

Comment:
Yikes. This isn't right.

`endSingleBatch` calls `moveForward`, which advances the `ifUnmodifiedSince` value to allow for the next upload. But that's a side-effect, and isn't part of the contract! You're just using that to leak the result of the upload. Don't do that — if you need the timestamp that was chained out of the upload, push it through the API.

`lastTimestamp` also typically won't be right. If you need to fall back to that, it means something went awry.